pymc_experimental/marginal_model.py

```diff
@@ -0,0 +1,440 @@
+    # This implies (?) that the dimensions are completely independent and a logp graph
+    # can ultimately be generated that is proportional to the support domain and not
+    # to the variables dimensions
+    # We don't need to worry about this if the  RV is scalar.
+    if np.prod(constant_fold(tuple(rv_to_marginalize.shape))) > 1:
+        if not is_elemwise_subgraph(rv_to_marginalize, dependent_rvs_input_rvs, dependent_rvs):
+            raise NotImplementedError(
+                "The subgraph between a marginalized RV and its dependents includes non Elemwise operations. "
+                "This is currently not supported",
+            )
+
+    input_rvs = [*marginalized_rv_input_rvs, *dependent_rvs_input_rvs]
+    rvs_to_marginalize = [rv_to_marginalize, *dependent_rvs]
+
+    outputs = rvs_to_marginalize
+    # Clone replace inner RV rng inputs so that we can be sure of the update order
+    # replace_inputs = {rng: rng.type() for rng in updates_rvs_to_marginalize.keys()}
+    # Clone replace outter RV inputs, so that their shared RNGs don't make it into
+    # the inner graph of the marginalized RVs
+    # FIXME: This shouldn't be needed!
+    replace_inputs = {}
+    replace_inputs.update({input_rv: input_rv.type() for input_rv in input_rvs})
+    cloned_outputs = clone_replace(outputs, replace=replace_inputs)
+
+    marginalization_op = FiniteDiscreteMarginalRV(
+        inputs=list(replace_inputs.values()),
+        outputs=cloned_outputs,
+        ndim_supp=0,
+    )
+    marginalized_rvs = marginalization_op(*replace_inputs.keys())
+    fgraph.replace_all(tuple(zip(rvs_to_marginalize, marginalized_rvs)))
+    return rvs_to_marginalize, marginalized_rvs
+
+
+@_get_measurable_outputs.register(FiniteDiscreteMarginalRV)
+def _get_measurable_outputs_finite_discrete_marginal_rv(op, node):
+    # Marginalized RVs are not measurable
+    return node.outputs[1:]
+
+
+def get_domain_of_finite_discrete_rv(rv: TensorVariable) -> Tuple[int, ...]:
+    op = rv.owner.op
+    if isinstance(op, Bernoulli):
+        return (0, 1)
+    elif isinstance(op, Categorical):
+        p_param = rv.owner.inputs[3]
+        return tuple(range(at.get_vector_length(p_param)))
+    elif isinstance(op, DiscreteUniform):
+        lower, upper = constant_fold(rv.owner.inputs[3:])
+        return tuple(range(lower, upper + 1))
+
+    raise NotImplementedError(f"Cannot compute domain for op {op}")
+
+
+@_logprob.register(FiniteDiscreteMarginalRV)
+def finite_discrete_marginal_rv_logp(op, values, *inputs, **kwargs):
+    marginalized_rvs_node = op.make_node(*inputs)
+    marginalized_rv, *dependent_rvs = clone_replace(
+        op.inner_outputs,
+        replace={u: v for u, v in zip(op.inner_inputs, marginalized_rvs_node.inputs)},
+    )
+
+    # Some inputs are not root inputs (such as transformed projections of value variables)
+    # Or cannot be used as inputs to an OpFromGraph (shared variables and constants)
+    inputs = list(inputvars(inputs))
+
```

Comment:
It might be worth explaining what lines below do.

Reply:
Definitely, I will add some

Reply:
I added some comments explaining what's going on there. Also docstrings to the class